pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentgenerationandpush/SegmentGenerationAndPushTaskGenerator.java

```diff
@@ -0,0 +1,342 @@
+      String batchSegmentIngestionType = IngestionConfigUtils.getBatchSegmentIngestionType(tableConfig);
+      BatchIngestionConfig batchIngestionConfig = tableConfig.getIngestionConfig().getBatchIngestionConfig();
+      List<Map<String, String>> batchConfigMaps = batchIngestionConfig.getBatchConfigMaps();
+      for (Map<String, String> batchConfigMap : batchConfigMaps) {
+        try {
+          URI inputDirURI =
+              SegmentGenerationUtils.getDirectoryURI(batchConfigMap.get(BatchConfigProperties.INPUT_DIR_URI));
+          updateRecordReaderConfigs(batchConfigMap);
+          List<SegmentZKMetadata> segmentsZKMetadata = Collections.emptyList();
+          // For append mode, we don't create segments for input file URIs already created.
+          if (BatchConfigProperties.SegmentIngestionType.APPEND.name().equalsIgnoreCase(batchSegmentIngestionType)) {
+            segmentsZKMetadata = _clusterInfoAccessor.getSegmentsZKMetadata(offlineTableName);
+          }
+          Set<String> existingSegmentInputFiles = getExistingSegmentInputFiles(segmentsZKMetadata);
+          Set<String> inputFilesFromRunningTasks = getInputFilesFromRunningTasks();
+          existingSegmentInputFiles.addAll(inputFilesFromRunningTasks);
+          LOGGER.info("Trying to extract input files from path: {}, "
+                  + "and exclude input files from existing segments metadata: {}, "
+                  + "and input files from running tasks: {}", inputDirURI, existingSegmentInputFiles,
+              inputFilesFromRunningTasks);
+          List<URI> inputFileURIs = getInputFilesFromDirectory(batchConfigMap, inputDirURI, existingSegmentInputFiles);
+          LOGGER.info("Final input files for task config generation: {}", inputFileURIs);
+          for (URI inputFileURI : inputFileURIs) {
+            Map<String, String> singleFileGenerationTaskConfig =
+                getSingleFileGenerationTaskConfig(offlineTableName, tableNumTasks, batchConfigMap, inputFileURI);
+            pinotTaskConfigs.add(new PinotTaskConfig(MinionConstants.SegmentGenerationAndPushTask.TASK_TYPE,
+                singleFileGenerationTaskConfig));
+            tableNumTasks++;
+
+            // Generate up to tableMaxNumTasks tasks each time for each table
+            if (tableNumTasks == tableMaxNumTasks) {
+              break;
+            }
+          }
+        } catch (Exception e) {
+          LOGGER.error("Unable to generate the SegmentGenerationAndPush task. [ table configs: {}, task configs: {} ]",
+              tableConfig, taskConfigs, e);
+        }
+      }
+    }
+    return pinotTaskConfigs;
+  }
+
+  private Set<String> getInputFilesFromRunningTasks() {
+    Set<String> inputFilesFromRunningTasks = new HashSet<>();
+    Map<String, TaskState> taskStates =
+        _clusterInfoAccessor.getTaskStates(MinionConstants.SegmentGenerationAndPushTask.TASK_TYPE);
+    for (String taskName : taskStates.keySet()) {
+      switch (taskStates.get(taskName)) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.